juju-tutorials/juju-tricks.md

```diff
@@ -0,0 +1,177 @@
+
+```console
+juju run --unit flannel/1 "relation-set -r cni:11 cni-conf-file=10-flannel.conflist"
+```
+
+### Triggering relation-changed hook
+
+The easiest way to trigger `relation-changed` hook is to change data in the relation using dummy variable that will be ignored by the other end of the relation. See chapter above, "Manually setting relation data" to see how to run `relation-set`.
+
+[note status="Reactive charms"]
+For the reactive charms, use 'juju run --unit foo/0 charms.reactive' to set/unset the flags and wait for the update-status hook to run
+[/note]
+
+## Debugging
+
+Duration: N/A
+
+## Inspecting state of reactive charms
+
+It's real simple to check states of the reactive charm. In the example below we will inspect states of the unit `flannel/1`
+
+```console
+juju run --unit flannel/1 'charms.reactive -p get_states'
+{'cni.configured': None,
+ 'cni.connected': None,
+ 'cni.is-master': None,
+ 'endpoint.cni.changed.cni-conf-file': None,
+ 'endpoint.cni.changed.egress-subnets': None,
+ 'endpoint.cni.changed.ingress-address': None,
+ 'endpoint.cni.changed.is_master': None,
+ 'endpoint.cni.changed.private-address': None,
+ 'endpoint.cni.joined': None,
+ 'etcd.available': {'conversations': ['reactive.conversations.etcd.global'],
+                    'relation': 'etcd'},
+ 'etcd.connected': {'conversations': ['reactive.conversations.etcd.global'],
+                    'relation': 'etcd'},
+ 'etcd.tls.available': {'conversations': ['reactive.conversations.etcd.global'],
+                        'relation': 'etcd'},
+ 'flannel.binaries.installed': None,
+ 'flannel.cni.available': None,
+ 'flannel.etcd.credentials.installed': None,
+ 'flannel.network.configured': None,
+ 'flannel.service.installed': None,
+ 'flannel.service.started': None,
+ 'flannel.version.set': None}
+```
+
+## Accessing juju internal database
+
+Duration: N/A
+
+[note status="Caution"]
+Changing values in the juju database can corrupt your juju installation. Proceed with caution and doublecheck before making any changes.
+[/note]
+
+Sometimes it may be necessary to access juju's database directly, although it should be considered as last resort solution. To log into the juju's database, you must first log into the juju controller machine.
+
+```console
+juju ssh -m controller 0
+```
+
+Once there, you can log into the juju's mongo database.
+
+```console
+mongo --sslAllowInvalidCertificates --authenticationDatabase admin --ssl -u $(sudo awk '/tag/ {print $2}' /var/lib/juju/agents/machine-?/agent.conf) -p $(sudo awk '/statepassword/ {print $2}' /var/lib/juju/agents/machine-?/agent.conf) localhost:37017/juju
+```
+
+### Manual change of Openstack endpoint certificates
+
+In case you are running juju controller on top of Openstack and for some reason the CA certificates on the endpoints change, you can't access the endpoint anymore with "certificate signed by unknown authority", the only way to fix it is to update the juju mongodb directly.
```

Comment:
"you are running a juju controller on top...", "the only way to fix it is to update Juju's mongodb directly"